### PR TITLE
Bug #3305 - Have embed.js load Popcorn via require

### PR DESCRIPTION
**Pull request by jbuck**
This pull request does a bunch of cool things. The primary goal was to have RequireJS load Popcorn, and as a side-effect we get:

1) Nuke the whole concept of external assets loaded by the server being different from the JS, eliminating a whole class of "which file do I edit to get my new plugin loaded"
2) IE9 shims will _always_ load before app code
3) We only load a single JS file now, so faster loading times
4) The r.js optimizer will concat all of the Popcorn files into `embed.js` so that we'll eventually be able to remove all of our build-related code for that

:yum: 

This is only the first part; I'll do a similar thing for butter.js in another patch.

### CODE REVIEW

File: cornfield/views/embed.jade [hunk 1]
```diff
@@ -2,6 +2,7 @@
 html(lang='en')
   head
     base(href='#{baseHref}')
+    script(src='/src/embed.js')
```
Comment (humphd) [inline cornfield/views/embed.jade]: Why move this?

File: src/embed.js [hunk 2]
```diff
@@ -242,21 +246,105 @@ function init( window, document ) {
...
+      "popcorn.loopPlugin": "../templates/assets/plugins/loopPlugin/popcorn.loopPlugin",
+      "popcorn.pausePlugin": "../templates/assets/plugins/pausePlugin/popcorn.pausePlugin",
+      "popcorn.popup": "../templates/assets/plugins/popup/popcorn.popup",
+      "popcorn.sequencer": "../templates/assets/plugins/sequencer/popcorn.sequencer",
+      "popcorn.skip": "../templates/assets/plugins/skip/popcorn.skip",
+      "popcorn.text": "../templates/assets/plugins/text/popcorn.text",
+      "popcorn.twitter": "../templates/assets/plugins/twitter/popcorn.twitter",
+      "popcorn.wikipedia": "../templates/assets/plugins/wikipedia/popcorn.wikipedia",
+
+      // Wrappers
```
Comment (humphd) [inline src/embed.js]: I'd move these after the ie8 shim above, since they are needed before players can ever be used.  It just helps to understand the deps a bit better--I realize the shim section below codifies it, but for reading purposes.

File: src/embed.js [hunk 3]
```diff
@@ -242,21 +246,105 @@ function init( window, document ) {
...
+      "popcorn.popup": [ "popcorn"],
+      "popcorn.sequencer": [ "popcorn", "popcorn.player" ],
+      "popcorn.skip": [ "popcorn" ],
+      "popcorn.text": [ "popcorn" ],
+      "popcorn.twitter": [ "popcorn" ],
+      "popcorn.wikipedia": [ "popcorn" ],
+
+      // Wrappers
+      "popcorn._MediaElementProto": [ "popcorn" ],
+      "popcorn.HTMLMediaElement": [ "popcorn" ],
```
Comment (humphd) [inline src/embed.js]: I think you should add `popcorn._MediaElementProto` here too, in case we do things with it down the road.  It won't hurt.

File: src/embed.js (hunk 2 above)
Comment (jbuck) [inline src/embed.js]: In the paths/shim config block, order doesn't matter; I just kept it the same as the shim config block. I'm happy to change the ordering of both, but I'd like both to match each other exactly because it'll be easier to add/remove files that way

File: cornfield/views/embed.jade (hunk 1 above)
Comment (jbuck) [inline cornfield/views/embed.jade]: Hm, I think I moved it originally as part of a different version of the patch, but it can probably go back to it's original location without harming anyone

File: src/embed.js [hunk 4]
```diff
@@ -408,15 +496,22 @@ function init( window, document ) {
...
+    rscript.onload = function() {
+      init();
+    };
+    rscript.src = "/external/require/require.js";
+    document.head.appendChild(rscript);
   } else {
-    init( window, window.document );
+    init();
   }
-}, false );
```
Comment (humphd) [inline src/embed.js]: Undo this change.

File: src/embed.js [hunk 5]
```diff
@@ -408,15 +496,22 @@ function init( window, document ) {
...
+  });
 }
 
 document.addEventListener( "DOMContentLoaded", function() {
   // Source tree case vs. require-built case.
   if ( typeof require === "undefined" ) {
-    Popcorn.getScript( "../../external/require/require.js", function() {
-      init( window, window.document );
-    });
+    var rscript = document.createElement("script");
```
Comment (humphd) [inline src/embed.js]: space around "script"

File: src/embed.js [hunk 6]
```diff
@@ -408,15 +496,22 @@ function init( window, document ) {
...
+    rscript.onload = function() {
+      init();
+    };
+    rscript.src = "/external/require/require.js";
+    document.head.appendChild(rscript);
```
Comment (humphd) [inline src/embed.js]: space around `rscript`

File: src/embed.js (hunk 3 above)
Comment (jbuck) [inline src/embed.js]: HTMLMediaElement uses the built-in audio/video elements, and doesn't depend on our MediaElementProto at all; if this file doesn't depend on it, why should it be listed as a dependency?

File: tools/build-embed.js [hunk 7]
```diff
@@ -7,22 +7,19 @@
...
-  },
-
   // Target the AMD loader shim as the main module to optimize,
   // so it shows up first in the built file,
   // since the embed modules use the define/require APIs that the almond
   // provides. Path is relative to baseUrl.
   name: '../tools/almond',
 
-  // Strip extra licenses
-  preserveLicenseComments: false,
```
Comment (humphd) [inline tools/build-embed.js]: We've ended up with a mess of license block comments in the final `embed.js`.  Why remove this?

File: src/embed.js (hunk 3 above)
Comment (humphd) [inline src/embed.js]: You're right.

File: cornfield/views/embed.jade (hunk 1 above)
Comment (humphd) [inline cornfield/views/embed.jade]: Yeah, please do.

File: src/embed.js (hunk 2 above)
Comment (humphd) [inline src/embed.js]: Yeah, let's do that.

Comment (Pomax) [inline src/embed.js]: why is this a hardcoded list, rather than an fs-grep for all the plugins, or a load-from-conf? The location is uniform, it's all in ../templates/assets/plugins

File: cornfield/views/embed.jade [hunk 8]
```diff
@@ -2,6 +2,7 @@
...
     !{templateScripts}
     !{externalAssets}
```
Comment (Pomax) [inline cornfield/views/embed.jade]: if we're doing requires for templates and assets in embed.js now, do we still need these two lines?

File: src/embed.js [hunk 9]
```diff
@@ -1,4 +1,8 @@
-function init( window, document ) {
+/*! This Source Code Form is subject to the terms of the MIT license
+ *  If a copy of the MIT license was not distributed with this file, you can
```
Comment (Pomax) [inline src/embed.js:2]: I know what "Source Code" means, but what's "Source Code Form"? Is just the indentation licensed and the actual code text isn't? O_o (can we change this to something that doesn't make people wonder what we're saying?)

File: cornfield/views/embed.jade (hunk 8 above)
Comment (jbuck) [inline cornfield/views/embed.jade]: Sadly yes; ripping this out will be another bug because it means that we can't have multiple-template publishing. Which I'd actually have no problem with removing! Just that it has large repercussions.

File: src/embed.js (hunk 9 above)
Comment (jbuck) [inline src/embed.js:2]: Feel free to file a bug on it! This is just a license copied from all of our other files

File: src/embed.js (hunk 2 above)
Comment (jbuck) [inline src/embed.js]: This code only gets executed in the browser, so no file system access. The locations are not uniform, we've got stuff in `templates/assets/` and various directories in `external/popcorn-js/`
